### PR TITLE
Adding exemplar support to go_client

### PR DESCRIPTION
**Pull request by arun-shopify**

### CODE REVIEW

File: prometheus/histogram.go [hunk 1]
```diff
@@ -573,6 +573,7 @@ type constHistogram struct {
 	sum        float64
 	buckets    map[float64]uint64
 	labelPairs []*dto.LabelPair
+	exemplars  []*dto.Exemplar
```
Comment (wperron) [inline prometheus/histogram.go:576]: Correct me if I'm wrong, this is necessary because `opentelemetry-collector-contrib`'s prometheus exporter uses the `NewConstHistogram` which doesn't support exemplar right? If that's the case, is this change going to be sufficient? Should it also include a change for `constHistogram` to implement the `ObserverWithExemplar` interface as well?

Comment (tanner-bruce) [reply prometheus/histogram.go:576]: I believe implementing `ObserverWithExemplar` would be incorrect, since this is a `ConstHistogram`, and not meant to be modified after creation

File: go.mod [hunk 2]
```diff
@@ -1,4 +1,4 @@
-module github.com/prometheus/client_golang
+module github.com/Shopify/client_golang
```
Comment (wperron) [inline go.mod:1]: I could be wrong, but iirc if we just `replace github.com/prometheus/client_golang => github.com/Shopify/client_golang main` it will resolve the module correctly _without_ having to change this line. You should double check though, I'm not 100% confident.

Comment (wperron) [reply go.mod:1]: if you do a replace in here to point `github.com/prometheus/client_golang => .` does that work? Also, I think the best case scenario here is that we get this upstream, so this would only be temporary anyway 🤷

Comment (arun-shopify) [reply go.mod:1]: thats a good idea, will try that out.

Comment (arun-shopify) [reply go.mod:1]: @wperron doing a replace of the package like that wont work for imports under `internal` folder: https://stackoverflow.com/questions/59342373/use-of-internal-package-not-allowed

Comment (tanner-bruce) [reply go.mod:1]: You should be able to leave all package names untouched in this repo, and then you do the `replace` in the repo where you use/import this one

File: prometheus/go_collector_go117.go [hunk 3]
```diff
@@ -22,9 +22,9 @@ import (
 	"runtime/metrics"
 	"sync"
 
+	"github.com/prometheus/client_golang/prometheus/internal"
```
Comment (wperron) [inline prometheus/go_collector_go117.go:25]: Pretty sure this is just an artifact from having changed to `Shopify` and reverted back, we can revert this change

**Review by wperron — APPROVED** [review]
Looks good overall, some minor nitpicks 💯

File: prometheus/histogram.go [hunk 4]
```diff
@@ -604,6 +611,13 @@ func (h *constHistogram) Write(out *dto.Metric) error {
 	return nil
 }
 
+func (h *constHistogram) GetExemplars() []*dto.Exemplar {
+	if h != nil {
+		return h.exemplars
+	}
+	return nil
+}
```
Comment (wperron) [inline prometheus/histogram.go:619]: Looking at [the PR in the collector contrib](https://github.com/Shopify/opentelemetry-collector-contrib/pull/1/files) this function doesn't seemed to be used, do we need to keep it in this patch?

File: prometheus/histogram.go [hunk 5]
```diff
@@ -596,6 +596,13 @@ func (h *constHistogram) Write(out *dto.Metric) error {
 	if len(buckets) > 0 {
 		sort.Sort(buckSort(buckets))
 	}
+
+	if len(h.exemplars) > 0 {
```
Comment (tanner-bruce) [inline prometheus/histogram.go:600]: This would panic if `len(h.exemplars) < len(buckets)`

Comment (arun-shopify) [reply prometheus/histogram.go:600]: good catch!

File: prometheus/histogram.go [hunk 6]
```diff
@@ -596,6 +596,19 @@ func (h *constHistogram) Write(out *dto.Metric) error {
...
+	//fmt.Println("exemplars: ", len(h.exemplars), " buckets: ", len(buckets))
+	//fmt.Println("bucket 0 exemplar: ", buckets[0].Exemplar.GetValue())
+	if len(h.exemplars) > 0 {
+		r := len(buckets)
+		if len(h.exemplars) < r {
+			r = len(h.exemplars)
+		}
+		for i := 0; i < r; i++ {
+			buckets[i].Exemplar = h.exemplars[i]
+		}
```
Comment (wperron) [inline prometheus/histogram.go:609]: In here, the exemplars could be out of order, so they could end up being added to the wrong bucket. There's an unexported function on the regular histogram type to find the right bucket for a given value, if you look under the hood you'll see it uses `sort.SearchFloat64s`, you could use that to find the correct bucket to add the exemplar to.

File: prometheus/histogram.go [hunk 7]
```diff
@@ -596,6 +596,22 @@ func (h *constHistogram) Write(out *dto.Metric) error {
 	if len(buckets) > 0 {
 		sort.Sort(buckSort(buckets))
 	}
+
+	if len(h.exemplars) > 0 {
+		r := len(buckets)
+		if len(h.exemplars) < r {
+			r = len(h.exemplars)
```
Comment (wperron) [inline prometheus/histogram.go:603]: You could do `r := min(len(buckets), len(h.exemplars))` here for legibility.

File: prometheus/histogram.go (hunk 6 above)
Comment (wperron) [reply prometheus/histogram.go:612]: I'm also looking at the `else` block just above. If we refactor this to use the `sort.SearchFloat64s` we could probably remove it, as well as the guard clause for the length and simply iterate over `h.exemplars`, find the right bucket to insert it to and do that. We can reduce the number of lines, improve clarity, and reduce the failures cases.

File: prometheus/histogram.go [hunk 8]
```diff
@@ -581,21 +582,38 @@ func (h *constHistogram) Desc() *Desc {
 
 func (h *constHistogram) Write(out *dto.Metric) error {
 	his := &dto.Histogram{}
+	// h.buckets, buckets and bounds are all the same length
 	buckets := make([]*dto.Bucket, 0, len(h.buckets))
+	bounds := make([]float64, 0)
```
Comment (fbogsany) [inline prometheus/histogram.go:587]:
```suggestion
	bounds := make([]float64, 0, len(h.buckets))
```

File: prometheus/histogram.go [hunk 9]
```diff
@@ -581,21 +582,38 @@ func (h *constHistogram) Desc() *Desc {
...
 			UpperBound:      proto.Float64(upperBound),
 		})
+		bounds = append(bounds, upperBound)
 	}
 
+	// make sure that both bounds and buckets have the same ordering
 	if len(buckets) > 0 {
 		sort.Sort(buckSort(buckets))
 	}
+	sort.Float64s(bounds)
```
Comment (fbogsany) [inline prometheus/histogram.go:603]:
```suggestion
		sort.Sort(buckSort(buckets))
		sort.Float64s(bounds)
	}
```

File: prometheus/histogram.go [hunk 10]
```diff
@@ -604,7 +604,8 @@ func (h *constHistogram) Write(out *dto.Metric) error {
 
 	if len(h.exemplars) > 0 {
 		r := len(buckets)
-		for i := 0; i < r; i++ {
+		l := len(h.exemplars)
+		for i := 0; i < r && i < l; i++ {
```
Comment (wperron) [inline prometheus/histogram.go:608]: This isn't strictly necessary, with the `sort.SearchFloat64s` change, those two lists can be different length and still work just fine.
